pkg/replicator/replicater.go

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package replicator
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/oklog/run"
+	"github.com/oklog/ulid"
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/thanos-io/thanos/pkg/compact"
+	"github.com/thanos-io/thanos/pkg/component"
+	"github.com/thanos-io/thanos/pkg/extflag"
+	"github.com/thanos-io/thanos/pkg/objstore/client"
+	"github.com/thanos-io/thanos/pkg/prober"
+	"github.com/thanos-io/thanos/pkg/runutil"
+	"github.com/thanos-io/thanos/pkg/server/http"
+)
+
+// ParseFlagMatchers parse flag into matchers.
+func ParseFlagMatchers(s []string) ([]*labels.Matcher, error) {
+	matchers := make([]*labels.Matcher, 0, len(s))
+
+	for _, l := range s {
+		parts := strings.SplitN(l, "=", 2)
+		if len(parts) != 2 {
+			return nil, errors.Errorf("unrecognized label %q", l)
+		}
+
+		labelName := parts[0]
+		if !model.LabelName.IsValid(model.LabelName(labelName)) {
+			return nil, errors.Errorf("unsupported format for label %s", l)
+		}
+
+		labelValue, err := strconv.Unquote(parts[1])
+		if err != nil {
+			return nil, errors.Wrap(err, "unquote label value")
+		}
+		newEqualMatcher, err := labels.NewMatcher(labels.MatchEqual, labelName, labelValue)
+		if err != nil {
+			return nil, errors.Wrap(err, "new equal matcher")
+		}
+		matchers = append(matchers, newEqualMatcher)
+	}
+
+	return matchers, nil
+}
+
+// RunReplicate replicate data based on config.
+func RunReplicate(
+	g *run.Group,
+	logger log.Logger,
+	reg *prometheus.Registry,
+	_ opentracing.Tracer,
+	httpMetricsBindAddr string,
+	labelSelector labels.Selector,
+	resolution compact.ResolutionLevel,
+	compaction int,
+	fromObjStoreConfig *extflag.PathOrContent,
+	toObjStoreConfig *extflag.PathOrContent,
+	singleRun bool,
+) error {
+	logger = log.With(logger, "component", "replicate")
+
+	level.Debug(logger).Log("msg", "setting up metric http listen-group")
+
+	// register a metrics http service.
+	httpProbe := prober.NewHTTP()
+	s := http.New(logger, reg, component.Replicate, httpProbe)
+
+	g.Add(func() error {
+		level.Info(logger).Log("msg", "Listening for metrics", "address", httpMetricsBindAddr)
```

Comment:
It's missing our probes (ready/healthy). Do you mind looking on another commands and copy similar here? (: